File: src/main/java/com/betts/IllegalGraphException.java
package com.betts;

/**
 * Thrown when connecting a node to itself in {@code Dag.Builder} (self-loop).
 */
public final class IllegalGraphException extends IllegalArgumentException {

  /** Creates an exception with the given detail message. */
  public IllegalGraphException(String message) {
    super(message);
  }

  /** Creates an exception with the given detail message and cause. */
  public IllegalGraphException(String message, Throwable cause) {
    super(message, cause);
  }
}

File: typescript/src/illegalGraphException.ts
/**
 * Thrown when connecting a node to itself in `Dag.Builder` (self-loop).
 */
export class IllegalGraphException extends Error {
  override readonly name = "IllegalGraphException";
  readonly cause?: unknown;

  /** Creates an exception with the given detail message. */
  constructor(message: string, options?: { cause?: unknown }) {
    super(message);
    this.cause = options?.cause;
  }
}
